core/src/test/java/com/google/cloud/sql/core/CloudSqlInstanceTest.java

```diff
@@ -309,6 +425,105 @@ public void testThatForceRefreshBalksWhenAForceRefreshIsInProgress() throws Exce
     assertThat(refreshCount.get()).isEqualTo(2);
   }
 
+  @Test
+  public void testRefreshRetriesOnAfterFailedAttempts() throws Exception {
```

Comment:
What lines of code are we covering in this test that we don't have covered above?

Reply:
This tests that `InstanceData getInstanceData(long timeoutMs)` will block until it has a valid InstanceData, 
even when currentInstanceData expires while a refresh attempt is in progress and subsequent refresh attempts fail. 

This is important for testing the way `handleRefreshResult()` chains futures together on failed 
refresh attempts. It also tests that `getInstanceData()` will block when the currentInstanceData has expired and a refresh attempt is in progress.